pandas 0.21.0 compatibility part 2

hdoupe opened the pull request:
This implements a cleaner solution to pandas 0.21.0 compatibility issue #1628 that was initially addressed by PR #1629. The approach here was advocated in this [comment](https://github.com/pandas-dev/pandas/issues/9422#issuecomment-342515489).  See pandas-dev/pandas#9422 for context.

MattHJensen commented:
This looks good to me. Thanks @hdoupe. Merging.

martinholmer commented:
@hdoupe,  I like the approach in #1634.  Thanks for the very good work.

hdoupe commented:
Thanks @martinholmer

martinholmer commented:
@hdoupe,
I've been looking more closely at the use in #1634 of the new `numpy.Series.zsum()` function.
You replaced `sum()` with `zsum()` everywhere in the `utilsprvt.py` file, which seems sensible.
But you didn't make that replacement in the `utils.py` file or in any other files where `sum()` is used.
Shouldn't we replace all the `sum()` functions with `zsum()` functions?

Please advise.  If we decide to do a complete replacement, I'll be happy to prepare the pull request.

hdoupe commented:
@martinholmer I thought about this some, too.  I decided not to do it because it seemed like a pretty invasive change.  However, there are some places where this should be used such as `utils.weighted`.  

The only downside that I can think of is that this might cover up a situation where an error should be thrown or the result really should be `NaN`.  I'm having trouble thinking of a case where this would be a problem.  The `zsum` function only changes the `sum` behavior when the dataframe is empty.

I'm not advocating this approach, but I was curious to see what would happen if we changed all of the sum operations to `zsum`.  So, I implemented a hacky solution on [this branch.](https://github.com/hdoupe/Tax-Calculator/tree/sum_experiment)  It still passes all of the tests.

In general, we should think really hard before we over-write changes made by a package like pandas.  So, my view is that this `zsum` function should be used sparingly.

martinholmer commented:
@hdoupe said:
>  I thought about this some, too. I decided not to do it because it seemed like a pretty invasive change. However, there are some places where this should be used such as utils.weighted.
>
> The only downside that I can think of is that this might cover up a situation where an error should be thrown or the result really should be NaN. I'm having trouble thinking of a case where this would be a problem. The zsum function only changes the sum behavior when the dataframe is empty.
>
> I'm not advocating this approach, but I was curious to see what would happen if we changed all of the sum operations to zsum. So, I implemented a hacky solution on this branch. It still passes all of the tests.

I agree we do not want to redefine the pandas Series method sum() as you did in your branch.

> In general, we should think really hard before we over-write changes made by a package like pandas. So, my view is that this zsum function should be used sparingly.

Will we just wait until something goes wrong before making any further use of zsum()?
Will we always know when something goes wrong?
Despite these questions, I'm comfortable going with the use-as-needed approach.

Except maybe you think zsum() should be used in weighted() in utils.py, is that right?
Should I make that change?  It's sort of weird that zsum() is defined in utils.py, but never used there.

**A question about the implementation of the zsum() function**:
pylint does not like the construct `len(self) > 0` saying it violates PEP8.
So, I've tentatively changed the code in #1637 to be this:
```
def zsum(self):
    """
    pandas 0.21.0 changes sum() behavior so that the result of applying sum
    over an empty Series is NaN.  Since we apply the sum() function over
    grouped DataFrames that may contain an empty Series, it makes more sense
    for us to have a sum() function that returns zero instead of NaN.
    """
    return self.sum() if self.size > 0 else 0
```
Does this make sense to you?

hdoupe commented:
@martinholmer said
> Will we just wait until something goes wrong before making any further use of zsum()?
> Will we always know when something goes wrong?

That's the downside of the as-needed approach.  It's difficult to know what situations will call for this.  I think any function that is heavily used after pandas `cut` or group-by type functions should probably use `zsum`.  This type of operation occurs frequently in `utils.py`.  So, I'm sure there are more places in `utils.py` where `zsum` should be used.

> Except maybe you think zsum() should be used in weighted() in utils.py, is that right?
> Should I make that change?

Sounds good. Go for it.

> It's sort of weird that zsum() is defined in utils.py, but never used there.

Right, `utilsprvt.py` isn't one of the modules that's imported in `__init__.py`.  I figured that I could put `zsum` in `utils.py` and not have to add another import to `__init__.py`.  However, I think `zsum` should only be used internally.  Thus, it belongs more in `utilsprvt.py`.  I guess we could add `from utilsprvt import zsum` to `__init__.py`.

Using `self.size` makes sense to me as long as `bool(len(self))` is equivalent to `bool(self.size)`
